server/src/main/java/org/elasticsearch/action/admin/cluster/state/TransportClusterStateAction.java

```diff
@@ -79,7 +79,10 @@ protected void masterOperation(Task task, final ClusterStateRequest request, fin
         if (acceptableClusterStatePredicate.test(state)) {
             ActionListener.completeWith(listener, () -> buildResponse(request, state));
         } else {
-            assert acceptableClusterStateOrFailedPredicate.test(state) == false;
+            // It is possible that the task is cancelled after the predicate has been executed, therefore we should take
+            // that into account as well for the assertion
+            assert acceptableClusterStateOrFailedPredicate.test(state) == false
+                || request.local() == false && cancellableTask.isCancelled();
```

Comment:
I was expecting the `cancellableTask.isCancelled()` here but not the `request.local()` check. Digging into why that's needed led me to conclude that there's a bug in #67413: local requests should also complete on the first cluster state update after cancellation.

Reply:
You're right, I've moved the cancellation check outside of `acceptableClusterStateOrFailedPredicate` making it deterministic given the same `ClusterState`. Now we check for cancellation in local requests too.